### ISSUE TITLE
Geom: Implement Shape hierarchy

### ISSUE THREAD

**Issue by scveloso** [UBC-Thunderbots/Software#958]
### Description of the task
Part of the geom redesign is implementing the following Shape hierarchy:

- Shape (i)
    - ConvexShape (i)
      - Circle
      - ConvexPolygon
   - Polygon
      - ConvexPolygon
         - Triangle
         - Rectangle
Notes: 
- ConvexPolygon inherits from both ConvexShape and Polygon
- Methods that use these belong in util and will be in a separate issue

### Acceptance criteria

- [ ] Implement the hierarchy as above
- [ ] Fields getters and setters for each class/interface

### Blocked By
- #956 Shapes depend on points / vectors(?)
- #957 Shapes depend on splines rays and polynomials

**Comment by garethellis0** [comment]
Just increasing the difficulty on this one a bit. We should probably doc the criteria for each difficulty level. Opened https://github.com/UBC-Thunderbots/Software/issues/1032 for myself to do that.
